src/analysis/normed_space/hahn_banach/separation.lean

```diff
@@ -0,0 +1,359 @@
+    simp only [f.map_neg, f.map_add, f.map_sub, hf₁] at this,
+    linarith },
+  refine ⟨f, Inf (f '' B), _, _⟩,
+  { suffices : f '' A ⊆ Iio (Inf (f '' B)),
+    { intros a ha,
+      apply this ⟨_, ha, rfl⟩ },
+    rw ←interior_Iic,
+    apply interior_maximal,
+    { rintro _ ⟨a, ha, rfl⟩,
+      apply le_cInf ⟨f b₀, _⟩,
+      { rintro _ ⟨b', hb, rfl⟩,
+        exact (forall_lt _ ha _ hb).le },
+      { exact mem_image_of_mem _ hb₀ } },
+    refine f.is_open_map _ _ hA₂,
+    rintro rfl,
+    simpa using hf₁ },
+  { intros b hb,
+    apply cInf_le ⟨f a₀, _⟩ (mem_image_of_mem _ hb),
+    rintro _ ⟨b', hb', rfl⟩,
+    exact (forall_lt _ ha₀ _ hb').le },
+end
+
+theorem geometric_hahn_banach_open_point {A : set E} {x : E} (hA₁ : convex ℝ A) (hA₂ : is_open A)
+  (disj : x ∉ A) :
+  ∃ (f : E →L[ℝ] ℝ), (∀ a ∈ A, f a < f x) :=
+let ⟨f, s, hA, hx⟩ := geometric_hahn_banach_open hA₁ hA₂ (convex_singleton x)
+  (disjoint_singleton_right.2 disj)
+  in ⟨f, λ a ha, lt_of_lt_of_le (hA a ha) (hx x (mem_singleton _))⟩
+
+theorem geometric_hahn_banach_point_open {x : E} {B : set E} (hB₁ : convex ℝ B) (hB₂ : is_open B)
+  (disj : x ∉ B) :
+  ∃ (f : E →L[ℝ] ℝ), (∀ b ∈ B, f x < f b) :=
+let ⟨f, hf⟩ := geometric_hahn_banach_open_point hB₁ hB₂ disj in ⟨-f, by simpa⟩
+
+theorem geometric_hahn_banach_open_open {A B : set E} (hA₁ : convex ℝ A) (hA₂ : is_open A)
+  (hB₁ : convex ℝ B) (hB₃ : is_open B) (disj : disjoint A B) :
+  ∃ (f : E →L[ℝ] ℝ) (s : ℝ), (∀ a ∈ A, f a < s) ∧ (∀ b ∈ B, s < f b) :=
+begin
+  obtain (rfl | ⟨a₀, ha₀⟩) := A.eq_empty_or_nonempty,
+  { exact ⟨0, -1, by simp, λ b hb, by norm_num⟩ },
+  obtain (rfl | ⟨b₀, hb₀⟩) := B.eq_empty_or_nonempty,
+  { exact ⟨0, 1, λ a ha, by norm_num, by simp⟩ },
+  obtain ⟨f, s, hf₁, hf₂⟩ := geometric_hahn_banach_open hA₁ hA₂ hB₁ disj,
+  have : f ≠ 0,
+  { rintro rfl,
+    exact (hf₁ _ ha₀).not_le (hf₂ _ hb₀) },
+  have : is_open_map f := f.is_open_map this,
+  refine ⟨f, s, hf₁, _⟩,
+  suffices : f '' B ⊆ Ioi s,
+  { exact λ b hb, this ⟨b, ‹_›, rfl⟩ },
+  rw ←interior_Ici,
+  refine interior_maximal _ (this _ hB₃),
+  rintro _ ⟨_, _, rfl⟩,
+  exact hf₂ _ ‹_›,
+end
+
+open filter
+open_locale topological_space
+
+/-- If `A`, `B` are disjoint convex sets, `A` is compact and `B` is closed then we can find open
+disjoint convex sets containing them. -/
+-- TODO: This proof uses the normed space structure of `E`, but it could work for locally convex
+-- topological vector spaces: instead of taking the balls around 0 with radius 1/n, we could show
+-- there must be some convex neighbourhood `W` of 0 which make `A + W` and `B + W` disjoint?
+theorem closed_compact_separate {A B : set E} (hA₁ : convex ℝ A) (hA₂ : is_compact A)
+  (hB₁ : convex ℝ B) (hB₃ : is_closed B) (disj : disjoint A B) :
+  ∃ U V, is_open U ∧ is_open V ∧ convex ℝ U ∧ convex ℝ V ∧ A ⊆ U ∧ B ⊆ V ∧ disjoint U V :=
+begin
+  have : ∃ (n : ℕ), disjoint (A + metric.ball 0 (n+1)⁻¹) (B + metric.ball 0 (n+1)⁻¹),
```

Comment:
Can you make that `have` into a lemma?
In general I still find some of the proofs in this PR a bit long, by which I mean that some of the steps of the proofs likely could be turned into auxiliary lemmas. That's what I tried to do in my last two suggestions.

Reply:
Yeah of course. I know how splitting up lemmas work. Truth is I'm been lazy on that one 😛. I was mostly happy with it being finally unblocked from prerequisites.